File: 032_Graph-Implementation.py
# Graph Implementation using Adjacency Matrix with Vertex Data

# 1. Basic Graph using Adjacency Matrix (Undirected, Unweighted)
vertexData = ['A', 'B', 'C', 'D']

adjacency_matrix = [
    [0, 1, 1, 1],  # Edges for A
    [1, 0, 1, 0],  # Edges for B
    [1, 1, 0, 0],  # Edges for C
    [1, 0, 0, 0]   # Edges for D
]

def print_adjacency_matrix(matrix):
    print("\nAdjacency Matrix:")
    for row in matrix:
        print(row)

print('vertexData:', vertexData)
print_adjacency_matrix(adjacency_matrix)

def print_connections(matrix, vertices):
    print("\nConnections for each vertex:")
    for i in range(len(vertices)):
        print(f"{vertices[i]}: ", end="")
        for j in range(len(vertices)):
            if matrix[i][j]:
                print(vertices[j], end=" ")
        print()

print_connections(adjacency_matrix, vertexData)

# 2. Graph Implementation Using Classes (Undirected, Unweighted)
class Graph:
    def __init__(self, size):
        self.adj_matrix = [[0] * size for _ in range(size)]
        self.size = size
        self.vertex_data = [''] * size

    def add_edge(self, u, v):
        if 0 <= u < self.size and 0 <= v < self.size:
            self.adj_matrix[u][v] = 1
            self.adj_matrix[v][u] = 1

    def add_vertex_data(self, vertex, data):
        if 0 <= vertex < self.size:
            self.vertex_data[vertex] = data

    def print_graph(self):
        print("Adjacency Matrix:")
        for row in self.adj_matrix:
            print(' '.join(map(str, row)))
        print("\nVertex Data:")
        for vertex, data in enumerate(self.vertex_data):
            print(f"Vertex {vertex}: {data}")

g = Graph(4)
g.add_vertex_data(0, 'A')
g.add_vertex_data(1, 'B')
g.add_vertex_data(2, 'C')
g.add_vertex_data(3, 'D')
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(0, 3)
g.add_edge(1, 2)

g.print_graph()

# 3. Directed and Weighted Graph Using Classes
class DirectedWeightedGraph:
    def __init__(self, size):
        self.adj_matrix = [[None] * size for _ in range(size)]
        self.size = size
        self.vertex_data = [''] * size

    def add_edge(self, u, v, weight):
        if 0 <= u < self.size and 0 <= v < self.size:
            self.adj_matrix[u][v] = weight

    def add_vertex_data(self, vertex, data):
        if 0 <= vertex < self.size:
            self.vertex_data[vertex] = data

    def print_graph(self):
        print("Adjacency Matrix:")
        for row in self.adj_matrix:
            print(' '.join(map(lambda x: str(x) if x is not None else '0', row)))
        print("\nVertex Data:")
        for vertex, data in enumerate(self.vertex_data):
            print(f"Vertex {vertex}: {data}")

dg = DirectedWeightedGraph(4)
dg.add_vertex_data(0, 'A')
dg.add_vertex_data(1, 'B')
dg.add_vertex_data(2, 'C')
dg.add_vertex_data(3, 'D')
dg.add_edge(0, 1, 3)  # A -> B

dg.add_edge(0, 2, 2)  # A -> C
dg.add_edge(3, 0, 4)  # D -> A
dg.add_edge(2, 1, 1)  # C -> B

dg.print_graph()

File: 032_Graph-Implementation.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
    int size;
    vector<vector<int>> adj_matrix;
    vector<char> vertices;

public:
    Graph(const vector<char>& vertexData) {
        vertices = vertexData;
        size = vertices.size();
        adj_matrix.resize(size, vector<int>(size, 0));
    }

    // Initialize the adjacency matrix directly
    void setAdjacencyMatrix(const vector<vector<int>>& matrix) {
        if (matrix.size() == size && matrix[0].size() == size) {
            adj_matrix = matrix;
        }
        else {
            cout << "Error: Matrix size mismatch\n";
        }
    }

    void printAdjacencyMatrix() {
        cout << "\nAdjacency Matrix:\n";
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                cout << adj_matrix[i][j] << " ";
            }
            cout << "\n";
        }
    }

    void printConnections() {
        cout << "\nConnections for each vertex:\n";
        for (int i = 0; i < size; i++) {
            cout << vertices[i] << ": ";
            for (int j = 0; j < size; j++) {
                if (adj_matrix[i][j] == 1) {
                    cout << vertices[j] << " ";
                }
            }
            cout << "\n";
        }
    }
};


int main() {
    vector<char> vertexData = {'A', 'B', 'C', 'D'};
    vector<vector<int>> adjacency_matrix = {
        {0, 1, 1, 1},
        {1, 0, 1, 0},
        {1, 1, 0, 0},
        {1, 0, 0, 0}
    };

    cout << "vertexData: ";
    for (char v : vertexData) cout << v << " ";
    cout << "\n";

    Graph g(vertexData);
    g.setAdjacencyMatrix(adjacency_matrix);

    g.printAdjacencyMatrix();
    g.printConnections();

    return 0;
}
